File: app/src/main/java/com/codinglikeapirate/pocitaj/SolveView.java
package com.codinglikeapirate.pocitaj;

import android.annotation.SuppressLint;
import android.content.Context;
import android.graphics.Bitmap;
import android.graphics.Canvas;
import android.graphics.Paint;
import android.graphics.Path;
import android.graphics.Typeface;
import android.text.TextPaint;
import android.util.AttributeSet;
import android.util.Log;
import android.util.TypedValue;
import android.view.MotionEvent;
import android.view.View;

import androidx.annotation.NonNull;

import com.codinglikeapirate.pocitaj.StrokeManager.ContentChangedListener;
import com.google.mlkit.vision.digitalink.Ink;

/**
 * Main view for rendering content.
 *
 * <p>The view accepts touch inputs, renders them on screen, and passes the content to the
 * StrokeManager. The view is also able to draw content from the StrokeManager.
 */
public class SolveView extends View implements ContentChangedListener {
  private static final String TAG = "SolveView";
  private static final int STROKE_WIDTH_DP = 3;

  private final Paint currentStrokePaint;
  private final Paint canvasPaint;

  private final TextPaint lastResultPaint;
  private final Path currentStroke;
  private final TextPaint largeTextPaint;
  private String lastResult = "";
  private String stats = "";
  private Canvas drawCanvas;
  private Bitmap canvasBitmap;
  private StrokeManager strokeManager;
  private ExerciseBook exerciseBook;
  private QuestionView questionView;

  public SolveView(Context context) {
    this(context, null);
  }

  public SolveView(Context context, AttributeSet attributeSet) {
    super(context, attributeSet);
    currentStrokePaint = new Paint();
    currentStrokePaint.setColor(0xFFFF00FF); // pink.
    currentStrokePaint.setAntiAlias(true);
    // Set stroke width based on display density.
    currentStrokePaint.setStrokeWidth(TypedValue.applyDimension(TypedValue.COMPLEX_UNIT_DIP, STROKE_WIDTH_DP, getResources().getDisplayMetrics()));
    currentStrokePaint.setStyle(Paint.Style.STROKE);
    currentStrokePaint.setStrokeJoin(Paint.Join.ROUND);
    currentStrokePaint.setStrokeCap(Paint.Cap.ROUND);

    currentStroke = new Path();
    canvasPaint = new Paint(Paint.DITHER_FLAG);

    lastResultPaint = new TextPaint();
    // green
    lastResultPaint.setColor(0xFF116611);
    lastResultPaint.setTextSize(80);

    largeTextPaint = new TextPaint();
    largeTextPaint.setColor(0xAAAAAAFF);
    largeTextPaint.setAntiAlias(true);
    largeTextPaint.setTypeface(Typeface.SERIF);
    largeTextPaint.setTextSize(250);
  }

  void setStrokeManager(@NonNull StrokeManager strokeManager) {
    this.strokeManager = strokeManager;
  }

  public void setExerciseBook(@NonNull ExerciseBook exerciseBook) {
    this.exerciseBook = exerciseBook;
  }

  @Override
  protected void onSizeChanged(int width, int height, int oldWidth, int oldHeight) {
    Log.i(TAG, "onSizeChanged");
    canvasBitmap = Bitmap.createBitmap(width, height, Bitmap.Config.ARGB_8888);
    drawCanvas = new Canvas(canvasBitmap);
    if (width != oldWidth || height != oldHeight) {
      redrawContent();
    }
    invalidate();
  }

  public void redrawContent() {
    clear();
    Ink currentInk = strokeManager.getCurrentInk();
    drawInk(currentInk, currentStrokePaint);

    drawQuestion();
    questionView.setText(exerciseBook.getLast().question());
    invalidate();
  }

  private void drawInk(@NonNull Ink ink, Paint paint) {
    for (Ink.Stroke s : ink.getStrokes()) {
      drawStroke(s, paint);
    }
  }

  private void drawStroke(@NonNull Ink.Stroke s, Paint paint) {
    Path path = new Path();
    for (Ink.Point p : s.getPoints()) {
      if (path.isEmpty()) {
        path.moveTo(p.getX(), p.getY());
      } else {
        path.lineTo(p.getX(), p.getY());
      }
    }
    drawCanvas.drawPath(path, paint);
  }

  private void drawQuestion() {
    drawCanvas.drawText(exerciseBook.getLast().question(), 100, 400, largeTextPaint);
  }

  public void clear() {
    currentStroke.reset();
    onSizeChanged(canvasBitmap.getWidth(), canvasBitmap.getHeight(), canvasBitmap.getWidth(), canvasBitmap.getHeight());
  }

  @Override
  protected void onDraw(Canvas canvas) {
    canvas.drawBitmap(canvasBitmap, 0, 0, canvasPaint);
    canvas.drawText(lastResult, 30, canvasBitmap.getHeight() - 100, lastResultPaint);
    canvas.drawText(stats, canvasBitmap.getWidth() - 450, canvasBitmap.getHeight() - 100, lastResultPaint);
    canvas.drawPath(currentStroke, currentStrokePaint);
  }

  @SuppressLint("ClickableViewAccessibility")
  @Override
  public boolean onTouchEvent(MotionEvent event) {
    int action = event.getActionMasked();
    float x = event.getX();
    float y = event.getY();

    switch (action) {
      case MotionEvent.ACTION_DOWN:
        currentStroke.moveTo(x, y);
        break;
      case MotionEvent.ACTION_MOVE:
        currentStroke.lineTo(x, y);
        break;
      case MotionEvent.ACTION_UP:
        currentStroke.lineTo(x, y);
        drawCanvas.drawPath(currentStroke, currentStrokePaint);
        currentStroke.reset();
        break;
      default:
        break;
    }
    if (!strokeManager.addNewTouchEvent(event)) {
      Log.w(TAG, "onTouchEvent: stroke manager didn't process event");
    }
    invalidate();
    return true;
  }

  @Override
  public void onContentChanged() {
    if (exerciseBook.getLast().solved()) {
      Log.e(TAG, "last was solved");
    }
    String text = strokeManager.getContent().get(strokeManager.getContent().size() - 1).text;
    int result = ExerciseBook.NOT_RECOGNIZED;
    try {
      result = Integer.parseInt(text);
    } catch (NumberFormatException ignored) {
    }

    boolean correct = exerciseBook.getLast().solve(result);
    lastResult = exerciseBook.getLast().equation();
    lastResultPaint.setColor(correct ? 0xFF33AA33 : 0xFFFF8888);
    stats = exerciseBook.getStats();

    // do animation
    if (result != ExerciseBook.NOT_RECOGNIZED) {
      exerciseBook.generate();
    }
    redrawContent();
  }

  public void setQuestionView(QuestionView questionView) {
    this.questionView = questionView;
  }
}

File: app/src/main/java/com/codinglikeapirate/pocitaj/SolveView.kt
package com.codinglikeapirate.pocitaj

import android.annotation.SuppressLint
import android.content.Context
import android.graphics.Bitmap
import android.graphics.Canvas
import android.graphics.Paint
import android.graphics.Path
import android.text.TextPaint
import android.util.AttributeSet
import android.util.Log
import android.util.TypedValue
import android.view.MotionEvent
import android.view.View
import android.widget.TextView
import androidx.core.content.ContextCompat
import com.codinglikeapirate.pocitaj.StrokeManager.ContentChangedListener
import com.google.mlkit.vision.digitalink.Ink

/**
 * Main view for rendering content.
 *
 * The view accepts touch inputs, renders them on screen, and passes the content to the
 * StrokeManager. The view is also able to draw content from the StrokeManager.
 */
class SolveView @JvmOverloads constructor(
    context: Context?,
    attrs: AttributeSet? = null
) :
    View(context, attrs), ContentChangedListener {

    private val currentStrokePaint = Paint().apply {
        color = ContextCompat.getColor(context!!, R.color.current_stroke_color)
        isAntiAlias = true
        strokeWidth = TypedValue.applyDimension(
            TypedValue.COMPLEX_UNIT_DIP,
            STROKE_WIDTH_DP,
            resources.displayMetrics
        )
        style = Paint.Style.STROKE
        strokeJoin = Paint.Join.ROUND
        strokeCap = Paint.Cap.ROUND
    }

    private val canvasPaint = Paint(Paint.DITHER_FLAG)
    private val lastResultPaint = TextPaint().apply {
        color = ContextCompat.getColor(context!!, R.color.correct_text_color)
        textSize = LAST_RESULT_TEXT_SIZE
    }

    private val currentStroke = Path()
    private var lastResult = ""
    private lateinit var drawCanvas: Canvas
    private lateinit var canvasBitmap: Bitmap
    private var strokeManager: StrokeManager? = null
    private lateinit var exerciseBook: ExerciseBook
    private var questionView: TextView? = null

    fun setStrokeManager(strokeManager: StrokeManager) {
        this.strokeManager = strokeManager
        strokeManager.expectedResult = exerciseBook.last.getExpectedResult()
    }

    fun setExerciseBook(exerciseBook: ExerciseBook) {
        this.exerciseBook = exerciseBook
    }

    override fun onSizeChanged(width: Int, height: Int, oldWidth: Int, oldHeight: Int) {
        Log.i(TAG, "onSizeChanged")
        canvasBitmap = Bitmap.createBitmap(width, height, Bitmap.Config.ARGB_8888)
        drawCanvas = Canvas(canvasBitmap)
        if (width != oldWidth || height != oldHeight) {
            redrawContent()
        }
        invalidate()
    }

    private fun redrawContent() {
        strokeManager ?: return
        clear()
        val currentInk = strokeManager!!.currentInk
        drawInk(currentInk, currentStrokePaint)

        drawQuestion()
        questionView?.text = exerciseBook.last.question()
        invalidate()
    }

    private fun drawInk(ink: Ink, paint: Paint) {
        ink.strokes.forEach { drawStroke(it, paint) }
    }

    private fun drawStroke(s: Ink.Stroke, paint: Paint) {
        val path = Path()
        s.points.forEach { p ->
            if (path.isEmpty) {
                path.moveTo(p.x, p.y)
            } else {
                path.lineTo(p.x, p.y)
            }
        }
        drawCanvas.drawPath(path, paint)
    }

    /** @noinspection EmptyMethod */
    private fun drawQuestion() {
        // Method is empty, no need to do anything here.
    }

    private fun clear() {
        currentStroke.reset()
        onSizeChanged(
            canvasBitmap.width,
            canvasBitmap.height,
            canvasBitmap.width,
            canvasBitmap.height
        )
    }

    override fun onDraw(canvas: Canvas) {
        canvas.drawBitmap(canvasBitmap, 0f, 0f, canvasPaint)
        canvas.drawText(lastResult, 30f, canvasBitmap.height - 100f, lastResultPaint)
        canvas.drawPath(currentStroke, currentStrokePaint)
    }

    @SuppressLint("ClickableViewAccessibility")
    override fun onTouchEvent(event: MotionEvent): Boolean {
        val x = event.x
        val y = event.y

        when (event.actionMasked) {
            MotionEvent.ACTION_DOWN -> {
                currentStroke.moveTo(x, y)
            }

            MotionEvent.ACTION_MOVE -> currentStroke.lineTo(x, y)
            MotionEvent.ACTION_UP -> {
                currentStroke.lineTo(x, y)
                drawCanvas.drawPath(currentStroke, currentStrokePaint)
                currentStroke.reset()
            }
        }
        if (strokeManager?.addNewTouchEvent(event) == false) {
            Log.w(TAG, "onTouchEvent: stroke manager didn't process event")
        }
        invalidate()
        return true
    }

    override fun onNewRecognizedText(text: String?, correct: Boolean) {
        // assume the exerciseBook was already updated - in the activity
        lastResult = exerciseBook.historyList.elementAtOrNull(exerciseBook.historyList.size - 2)?.equation() ?: ""
        lastResultPaint.color = ContextCompat.getColor(
            context,
            if (correct) R.color.correct_text_color else R.color.incorrect_text_color
        )
        redrawContent()
    }

    override fun onMisparsedRecognizedText(text: String?) {
        redrawContent()
    }

    fun setQuestionView(questionView: TextView) {
        this.questionView = questionView
    }

    companion object {
        private const val TAG = "SolveView"
        private const val STROKE_WIDTH_DP = 6f
        private const val LAST_RESULT_TEXT_SIZE = 80f
    }
}